client/components/site-selector-modal/index.jsx

```diff
@@ -35,45 +37,46 @@ class SiteSelectorModal extends Component {
 		// className: class name(s) to be added to the Dialog
 		className: PropTypes.string,
 		// from localize()
-		translate: PropTypes.func.isRequired
+		translate: PropTypes.func.isRequired,
+		// connected props
+		primarySiteId: PropTypes.number
 	}
 
 	constructor( props ) {
 		super( props );
 
-		const primarySite = sitesList.getPrimary();
-		let filteredSites = sitesList.getVisible();
+		const { primarySiteId } = this.props;
+		let filteredSiteIds = sitesList.getVisible().map( ( site ) => site.ID );
 
 		if ( props.filter ) {
-			filteredSites = filteredSites.filter( props.filter );
+			filteredSiteIds = filteredSiteIds.filter( props.filter );
 		}
 
 		this.state = {
-			site: includes( filteredSites, primarySite )
-				? primarySite
-				: filteredSites[ 0 ]
+			siteId: includes( filteredSiteIds, primarySiteId )
```

Comment:
I was thinking of something to abstract the question of if `primarySite` belongs to the user and is visible. The actual implementation details are less important, even if we are doing something funky like filtering through the entire list.